### PR TITLE
Adding helper functions for switching of devtree backend in the same process

### PR DESCRIPTION
**Pull request by Swarnendu-R-C**
Context: It is now required that the devtree backend in BMC systems should be able to switch in the same process on runtime e.g.; from KERNEL to SBEFIFO or vice-versa. This is to support running extract SBE RC hardware procedure. This commit introduces the necessary changes in order to achieve that.

Solution: It basically introduces few functions which will clear the pdbg_dt_root and its associated children subsequently if the pdbg_targets_init function is called more than once now (earlier it was not allowed to call pdbg_targets_init function twice in the same process)

Test-1: The switching of the backend in the same process was tested through a standalone application which runs in a loop for multiple times and each time it sets the intial backend to SBEFIFO first and then switched it to KERNEL. Trace log is attached along with the PR

Test-2: While booting BMC up injected a clock error dueing IPL in between istep0 to istep2 and call the SBR RC extract procedure to get the error code after switching the backend from SBEFIFO to KERNEL. Once the extraction is done switched back to SBEFIFO for further processing. Journal log is attached with the PR.

Signed-off-by: swarnendu.roy.chowdhury@ibm.com

[journalLog.txt](https://github.com/open-power/pdbg/files/11597060/journalLog.txt)
[TraceLog_KARNEL_First.txt](https://github.com/open-power/pdbg/files/11597064/TraceLog_KARNEL_First.txt)
[TraceLog_SBEFIFO_First.txt](https://github.com/open-power/pdbg/files/11597066/TraceLog_SBEFIFO_First.txt)
Screenshot 2023-06-23 at 7 39 30 PM
Screenshot 2023-06-23 at 7 39 54 PM

### CODE REVIEW

File: libpdbg/target.h [hunk 1]
```diff
@@ -83,6 +83,18 @@ struct pdbg_target_class *require_target_class(const char *name);
 struct pdbg_target_class *get_target_class(struct pdbg_target *target);
 bool pdbg_target_is_class(struct pdbg_target *target, const char *class);
 
+/**
```
Comment (devenrao) [inline libpdbg/target.h:86]: why this isn;t static void function in device.c

Comment (Swarnendu-R-C) [reply libpdbg/target.h:86]: I don't know. Not my changes. My changes are only at LN87

File: libpdbg/device.c [hunk 2]
```diff
@@ -701,6 +701,73 @@ static void pdbg_targets_init_virtual(struct pdbg_target *node, struct pdbg_targ
...
+		return;
+
+	struct pdbg_target_class* target_class = get_target_class(target);
+	if (target_class)
+		list_del_from(&target_class->targets, &target->class_link);
+}
+
+/**
+ * @brief A function to delete/release an existing pdbg_target object along
+ * 		  with its children from the node list/dev tree. Should be only called
```
Comment (dhruvibm) [inline libpdbg/device.c:728]: keep in 80 lines

Comment (Swarnendu-R-C) [reply libpdbg/device.c]: Done

File: libpdbg/device.c [hunk 3]
```diff
@@ -701,6 +701,73 @@ static void pdbg_targets_init_virtual(struct pdbg_target *node, struct pdbg_targ
...
+}
+
+void pdbg_release_dt_root()
+{
+	if (pdbg_dt_root)
+	{	
+		pdbg_release_target(pdbg_dt_root);
+		if (pdbg_dt_root)
+			pdbg_dt_root = NULL;
+		//Reset the phandle count to zero
```
Comment (deepakala-k) [inline libpdbg/device.c:764]: Do we want to stick with c style comments?

Comment (Swarnendu-R-C) [reply libpdbg/device.c:764]: Yes, as it is a C library in every sense.

File: libpdbg/libpdbg.h [hunk 4]
```diff
@@ -1459,8 +1459,8 @@ bool pdbg_context_short(void);
  * switching the backend in the same process. It clears/releases the 
  * existing dev tree (if any) children by children along with the root node.
  * 
- * Call this function very judiciously before ##pdbg_set_backend() if there is a
```
Comment (deepakala-k) [inline libpdbg/libpdbg.h]: Same as above commit

File: src/tests/libpdbg_release_dt_root_test.c [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+    assert(count == 64);
+}
+
+static void testSetTwo(struct pdbg_target *root)
+{
+    struct pdbg_target *parent, *target = NULL;
+    pdbg_for_each_child_target(root, parent) 
+    {
+		const char *name = pdbg_target_dn_name(parent);
+		assert(strncmp(name, "proc", 4) == 0);
```
Comment (deepakala-k) [inline src/tests/libpdbg_release_dt_root_test.c:78]: Please fix the indentation issues.

File: libpdbg/dtb.c [hunk 6]
```diff
@@ -418,12 +418,6 @@ bool pdbg_set_backend(enum pdbg_backend backend, const char *backend_option)
 		return false;
 	}
 
-	if (pdbg_backend == backend)
```
[the remark this answers is not in the document]
Comment (Swarnendu-R-C) [reply libpdbg/dtb.c]: Yes, you are correct. It is removed to make the unit test cases run effectively (although logically also it was a dumb statement). Thus it is a part of unit test commit.

File: libpdbg/libpdbg.h (hunk 4 above)
Comment (Swarnendu-R-C) [reply libpdbg/libpdbg.h]: This was also changed during writing unit test cases.

File: src/tests/libpdbg_release_dt_root_test.c (hunk 5 above)
Comment (Swarnendu-R-C) [reply src/tests/libpdbg_release_dt_root_test.c]: Fixed

**Review by deepakala-k — APPROVED** [review]

File: libpdbg/device.c [hunk 7]
```diff
@@ -701,6 +701,73 @@ static void pdbg_targets_init_virtual(struct pdbg_target *node, struct pdbg_targ
 	}
 }
 
+/**
```
Comment (devenrao) [inline libpdbg/device.c:704]: fix the commit message it is > 55 characters or so

File: libpdbg/device.c [hunk 8]
```diff
@@ -701,6 +701,73 @@ static void pdbg_targets_init_virtual(struct pdbg_target *node, struct pdbg_targ
...
+ * @brief It releases/deletes a target from its target_class list.
```
Comment (devenrao) [inline libpdbg/device.c:705]: it's

File: libpdbg/device.c [hunk 9]
```diff
@@ -701,6 +701,73 @@ static void pdbg_targets_init_virtual(struct pdbg_target *node, struct pdbg_targ
...
+	struct pdbg_target *childTarget, *next = NULL;
+	list_for_each_safe(&target->children, childTarget, next, list)
+		pdbg_release_target(childTarget);
+		
+	if (target->class)
+			pdbg_del_target_from_target_class_list(target);
+
+	struct pdbg_target* parentTarget = target->parent;
+	if (parentTarget)
+		list_del_from(&parentTarget->children, &target->list);
```
Comment (devenrao) [inline libpdbg/device.c:750]: specify why we are doing this in the method description

**Review by devenrao — COMMENTED** [review]
minor comments looks good

File: src/tests/libpdbg_release_dt_root_test.c [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <inttypes.h>
+#include <assert.h>
+
+#include <libpdbg.h>
+
```
Comment (devenrao) [inline src/tests/libpdbg_release_dt_root_test.c:23]: the test cases are not related to the functionality to switch back-ends it is more of adding additional test cases if so it should be in another branch and another PR and it should hav its own test file not part of release_dt_root_test.c

here we should have test cases only for releasing dt_root

File: src/tests/libpdbg_release_dt_root_test.c [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+    struct pdbg_target *target = NULL;
+    int cnt = 0;
+
+    pdbg_for_each_target(classname, parent, target)
+        ++cnt;
+
+    return cnt;
+}
+
+static int count_class_target(const char *classname)
```
Comment (devenrao) [inline src/tests/libpdbg_release_dt_root_test.c:35]: this file should just have init device tree with fake DB, get the count of targets, now clear the targets and then if possible check the count again like using count_class_target, or else you can define a method where it gives count of targets or classes and you can validate here afte clearing the targets

File: libpdbg/device.c (hunk 7 above)
Comment (Swarnendu-R-C) [reply libpdbg/device.c:704]: Corrected

File: libpdbg/device.c (hunk 8 above)
Comment (Swarnendu-R-C) [reply libpdbg/device.c:705]: I think here it's is not needed as I am putting a s with "delete" and "release" verbs

File: libpdbg/device.c (hunk 9 above)
Comment (Swarnendu-R-C) [reply libpdbg/device.c:751]: Added

File: libpdbg/dtb.c [hunk 12]
```diff
@@ -414,14 +414,7 @@ bool pdbg_set_backend(enum pdbg_backend backend, const char *backend_option)
...
 	{
-		pdbg_log(PDBG_ERROR, "pdbg_set_backend() must be called before pdbg_targets_init() or 
-					after calling pdbg_release_dt_root() if a dev tree is already set\n");
-		return false;
-	}
-
-	if (pdbg_backend == backend)
-	{
-		pdbg_log(PDBG_ERROR, "New backend is same as the current backend. Not proceeding further\n");
+		pdbg_log(PDBG_ERROR, "pdbg_set_backend() must be called before pdbg_targets_init() or after calling pdbg_release_dt_root() if a dev tree is already set\n");
```
[the remark this answers is not in the document]
Comment (Swarnendu-R-C) [reply libpdbg/dtb.c]: Changed

File: src/tests/libpdbg_release_dt_root_test.c (hunk 10 above)
Comment (Swarnendu-R-C) [reply src/tests/libpdbg_release_dt_root_test.c:23]: Corrected as suggested. Now the test cases test for the release and then switching of backend also

File: src/tests/libpdbg_release_dt_root_test.c (hunk 11 above)
Comment (Swarnendu-R-C) [reply src/tests/libpdbg_release_dt_root_test.c:35]: Changed as suggested

File: src/tests/libpdbg_release_dt_root_test.c [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+    }
+}
+
+static void testSetOne(struct pdbg_target *root, const bool dtreeReleased)
+{
+    if (dtreeReleased)
+    {
+        assert(!root);
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (Swarnendu-R-C) [reply src/tests/libpdbg_release_dt_root_test.c]: Removed

**Comment by Swarnendu-R-C** [comment]
Opened the same in a new branch https://github.com/open-power/pdbg/pull/76 so closing this as not needed anymore
